File: python/day_01/main.py
import operator


class Elf:
    def __init__(self, raw_data: str) -> None:
        self.parse_data(raw_data)
        self.calculate_total_calories()

    def parse_data(self, raw_data: str) -> None:
        # Split on line breaks to seperate each elf calorie amount
        self.calories_list = map(int, raw_data.splitlines())

    def calculate_total_calories(self) -> None:
        self.total_calories = sum(self.calories_list)


class Elves:
    def __init__(self, input_data: str) -> None:
        self.create_elves(input_data)
        self.order_elves()

    def create_elves(self, input_data: str) -> None:
        # Split on blank lines to seperate each elf
        self.elves = map(Elf, input_data.split("\n\n"))

    def order_elves(self) -> None:
        order_key = operator.attrgetter("total_calories")
        self.ordered_elves = sorted(self.elves, key=order_key, reverse=True)

    def get_top_elf_calories(self) -> int:
        return self.ordered_elves[0].total_calories

    def get_top_n_elves_calories_sum(self, num_elves: int) -> int:
        top_n_elves = self.ordered_elves[:num_elves]
        return sum(elf.total_calories for elf in top_n_elves)


def main(input_data: str) -> tuple[int, int]:

    elves = Elves(input_data)

    part_1_answer = elves.get_top_elf_calories()
    part_2_answer = elves.get_top_n_elves_calories_sum(num_elves=3)

    return part_1_answer, part_2_answer

File: rust/src/day_01/main.rs
struct Elf {
    calories_list: Vec<i32>,
    total_calories: i32,
}

impl Elf {
    fn new(raw_data: &str) -> Elf {
        let mut elf = Elf {
            calories_list: Vec::new(),
            total_calories: 0,
        };

        elf.parse_data(raw_data);
        elf.calculate_total_calories();
        elf
    }

    fn parse_data(&mut self, raw_data: &str) {
        // Split on line breaks to seperate each elf calorie amount
        self.calories_list = raw_data
            .lines()
            .map(|calories| calories.parse::<i32>().unwrap())
            .collect();
    }

    fn calculate_total_calories(&mut self) {
        self.total_calories = self.calories_list.iter().sum();
    }
}

struct Elves {
    elves: Vec<Elf>,
}

impl Elves {
    fn new(input_data: String) -> Elves {
        let mut elves = Elves { elves: Vec::new() };

        elves.create_elves(input_data);
        elves.order_elves();
        elves
    }

    fn create_elves(&mut self, input_data: String) {
        // Split on blank lines to seperate each elf
        self.elves = input_data
            .split("\n\n")
            .map(|elf_raw_data| Elf::new(elf_raw_data))
            .collect();
    }

    fn order_elves(&mut self) {
        self.elves
            .sort_by(|a, b| b.total_calories.cmp(&a.total_calories));
    }

    fn get_top_elf_calories(&self) -> i32 {
        self.elves[0].total_calories
    }

    fn get_top_n_elves_calories_sum(&self, num_elves: usize) -> i32 {
        self.elves
            .iter()
            .take(num_elves)
            .map(|elf| elf.total_calories)
            .sum()
    }
}

pub fn main(input_data: String) -> (i32, i32) {
    let elves = Elves::new(input_data);

    let part_1_answer = elves.get_top_elf_calories();
    let part_2_answer = elves.get_top_n_elves_calories_sum(3);

    (part_1_answer, part_2_answer)
}
